python/dlisio/plumbing/basicobject.py

```diff
@@ -126,13 +134,48 @@ def strip(d):
         strip(self.__dict__)
         strip(self.stash)
 
-    def link(self, objects, sets):
+    def link(self, objects):
         """ Link objects
 
-        The default implementation is a no-op - individual object types
-        themselves are aware on how to link to other objects
+        Iterate through values in refs and link objects based on
+        provided linkage
         """
-        pass
+
+        linkage = self.linkage
+
+        for attr, link in linkage.items():
+            try:
+                ref = self.refs[attr]
+            except KeyError:
+                continue
+
+            if ref is None: continue
+
+            def get_link(v):
+                try:
+                    fp = link(v)
+                except (AttributeError, TypeError):
```

Comment:
Is AttributeError needed?

Reply:
Also, add  msg to logging.warning! This should only happen if someone adds a none obname/objref to self.refs before link/relink.